{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_MAG_and_divide(dir_genomes, MAG_TO_VIEW):\n",
    "\n",
    "    contigs = {}\n",
    "    contigs_rev = {}\n",
    "\n",
    "    pre = \"ACTG\"\n",
    "    post = \"TGAC\"\n",
    "\n",
    "    os.chdir(dir_genomes)\n",
    "    with open(MAG_TO_VIEW) as f: #### SELECT SPECIFIC MAG\n",
    "        v = f.readlines()\n",
    "        f.seek(0)\n",
    "        start = -1\n",
    "        curr_h = \"\"\n",
    "        curr_f = \"\"\n",
    "        for line in v:\n",
    "            if line.startswith(\">\"):\n",
    "                contigs[curr_h] = curr_f\n",
    "\n",
    "                curr_f = \"\"\n",
    "                curr_h = line.strip()\n",
    "            else:\n",
    "                curr_f += line.strip()\n",
    "        contigs[curr_h] = curr_f\n",
    "\n",
    "        start = -1\n",
    "        curr_h = \"\"\n",
    "        curr_f = \"\"\n",
    "        curr_f_rev = \"\"\n",
    "        for line in v:\n",
    "            if line.startswith(\">\"):\n",
    "                compl = curr_f.maketrans(pre,post)\n",
    "                fasta_compl = curr_f.translate(compl)\n",
    "                curr_f_rev = fasta_compl[::-1]\n",
    "                contigs_rev[curr_h] = curr_f_rev\n",
    "                curr_f = \"\"\n",
    "                curr_h = line.strip()\n",
    "            else:\n",
    "                curr_f += line.strip()\n",
    "\n",
    "        compl = curr_f.maketrans(pre,post)\n",
    "        fasta_compl = curr_f.translate(compl)\n",
    "        curr_f_rev = fasta_compl[::-1]\n",
    "        contigs_rev[curr_h] = curr_f_rev\n",
    "    \n",
    "    contigs.pop(\"\")\n",
    "    contigs_rev.pop(\"\")\n",
    "    \n",
    "    return contigs, contigs_rev\n",
    "\n",
    "def rewrite_MAG_removing_genes(MAG_TO_REWRITE, contigs, dir_output, DELETION_SIZE, dict_of_genes_to_remove):\n",
    "    '''\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    os.chdir(dir_output)\n",
    "\n",
    "    if not MAG_TO_REWRITE in os.listdir():\n",
    "        g = open(MAG_TO_REWRITE, \"a\")\n",
    "        g.close()\n",
    "\n",
    "    f = open(MAG_TO_REWRITE, \"w\")\n",
    "    for fasta, seq in contigs.items():\n",
    "        seq_to_write_by_lines = \"\"\n",
    "        f.write(fasta+\"\\n\")\n",
    "\n",
    "        if fasta in dict_of_genes_to_remove.keys():\n",
    "            \n",
    "            print(fasta) # debug\n",
    "\n",
    "            START = int(dict_of_genes_to_remove[fasta][0])\n",
    "            STRAND = dict_of_genes_to_remove[fasta][1]\n",
    "            GENE_TO_REMOVE = dict_of_genes_to_remove[fasta][2]\n",
    "            GENE_LENGHT = len(GENE_TO_REMOVE)\n",
    "\n",
    "            if DELETION_SIZE == \"small\":\n",
    "\n",
    "                PORTION_TO_REMOVE = round(GENE_LENGHT/100*30)\n",
    "                PORTION_TO_REMOVE_MINUS = GENE_LENGHT - PORTION_TO_REMOVE\n",
    "\n",
    "                if STRAND == \"+\":\n",
    "\n",
    "                    gene_seq_removed = GENE_TO_REMOVE[:PORTION_TO_REMOVE]\n",
    "                    gene_remove_index = seq.index(gene_seq_removed)\n",
    "\n",
    "                    gene_seq_maintained = GENE_TO_REMOVE[PORTION_TO_REMOVE:]\n",
    "                    #print(gene_seq_removed) #debug\n",
    "                    #print(gene_seq_maintained) #debug\n",
    "\n",
    "                    seq_to_write = \"\"\n",
    "                    seq_to_write += seq[:gene_remove_index]\n",
    "                    seq_to_write += seq[gene_remove_index+PORTION_TO_REMOVE:]\n",
    "\n",
    "                if STRAND == \"-\":\n",
    "                    gene_seq_removed = GENE_TO_REMOVE[PORTION_TO_REMOVE_MINUS:]\n",
    "                    gene_remove_index = seq.index(gene_seq_removed)\n",
    "                    gene_seq_maintained = GENE_TO_REMOVE[:PORTION_TO_REMOVE_MINUS]\n",
    "                    #print(gene_seq_removed) #debug\n",
    "                    #print(gene_seq_maintained) #debug\n",
    "                    #print(seq[(START+GENE_LENGHT):]) # debug\n",
    "\n",
    "                    seq_to_write = \"\"\n",
    "                    seq_to_write += seq[:gene_remove_index]\n",
    "                    seq_to_write += seq[START+GENE_LENGHT:]\n",
    "\n",
    "            elif DELETION_SIZE == \"big\":\n",
    "\n",
    "                PORTION_TO_REMOVE = round(GENE_LENGHT/100*70)\n",
    "                PORTION_TO_REMOVE_MINUS = GENE_LENGHT - PORTION_TO_REMOVE\n",
    "\n",
    "                if STRAND == \"+\":\n",
    "                    gene_seq_removed = GENE_TO_REMOVE[:PORTION_TO_REMOVE]\n",
    "                    gene_remove_index = seq.index(gene_seq_removed)\n",
    "                    gene_seq_maintained = GENE_TO_REMOVE[PORTION_TO_REMOVE:]\n",
    "                    #print(gene_seq_removed) #debug\n",
    "                    #print(gene_seq_maintained) #debug\n",
    "\n",
    "                    seq_to_write = \"\"\n",
    "                    seq_to_write += seq[:gene_remove_index]\n",
    "                    seq_to_write += seq[gene_remove_index+PORTION_TO_REMOVE:]\n",
    "\n",
    "                if STRAND == \"-\":\n",
    "\n",
    "                    gene_seq_removed = GENE_TO_REMOVE[PORTION_TO_REMOVE_MINUS:]\n",
    "                    gene_remove_index = seq.index(gene_seq_removed)\n",
    "\n",
    "                    gene_seq_maintained = GENE_TO_REMOVE[:PORTION_TO_REMOVE_MINUS]\n",
    "                    #print(gene_seq_removed) #debug\n",
    "                    #print(gene_seq_maintained) #debug\n",
    "\n",
    "                    seq_to_write = \"\"\n",
    "                    seq_to_write += seq[:gene_remove_index]\n",
    "                    seq_to_write += seq[START+GENE_LENGHT:]\n",
    "\n",
    "            for nt in seq_to_write:\n",
    "                if len(seq_to_write_by_lines) == 60:\n",
    "                    f.write(seq_to_write_by_lines+\"\\n\")\n",
    "                    seq_to_write_by_lines = \"\"\n",
    "                    pass\n",
    "                seq_to_write_by_lines += nt\n",
    "            f.write(seq_to_write_by_lines+\"\\n\")\n",
    "\n",
    "        else:\n",
    "            for nt in seq:\n",
    "                if len(seq_to_write_by_lines) == 60:\n",
    "                    f.write(seq_to_write_by_lines+\"\\n\")\n",
    "                    seq_to_write_by_lines = \"\"\n",
    "                    pass\n",
    "                seq_to_write_by_lines += nt\n",
    "            f.write(seq_to_write_by_lines+\"\\n\")\n",
    "    f.close()\n",
    "\n",
    "def rewrite_Genome_removing_genes(GENOME_TO_REWRITE, contigs, dir_output, DELETION_SIZE, dict_of_genes_to_remove):\n",
    "    '''\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    os.chdir(dir_output)\n",
    "    if not GENOME_TO_REWRITE in os.listdir():\n",
    "        g = open(GENOME_TO_REWRITE, \"a\")\n",
    "        g.close()\n",
    "\n",
    "    actual_position = -1\n",
    "\n",
    "    f = open(GENOME_TO_REWRITE, \"w\")\n",
    "    fasta = \">\"+GENOME_TO_REWRITE.replace(\".fna\", \"\").replace(\".fasta\", \"\")\n",
    "    seq = contigs[fasta]\n",
    "    seq_to_write_by_lines = \"\"\n",
    "    f.write(fasta+\"\\n\")\n",
    "    gene_positions = []\n",
    "\n",
    "    for gene_pos in dict_of_genes_to_remove.keys():\n",
    "        START = gene_pos\n",
    "        gene_positions.append(START)\n",
    "    #print(gene_positions) #debug\n",
    "\n",
    "    seq_to_write = \"\"\n",
    "    pos = -1\n",
    "    n_nt = -1\n",
    "\n",
    "    for nt in seq[:-1]:\n",
    "        #print(nt)\n",
    "        n_nt += 1\n",
    "        if not n_nt > pos:\n",
    "            continue\n",
    "        pos += 1\n",
    "        \n",
    "        #try:\n",
    "        if not str(pos) in gene_positions:\n",
    "            #print(seq[actual_position])\n",
    "            seq_to_write += seq[pos]\n",
    "\n",
    "        elif str(pos) in gene_positions:\n",
    "            GENE_TO_REMOVE = dict_of_genes_to_remove[str(pos)][1]\n",
    "            GENE_LENGHT = len(GENE_TO_REMOVE)\n",
    "            STRAND = dict_of_genes_to_remove[str(pos)][0]\n",
    "            #print(\"gene_len: \"+str(GENE_LENGHT)) # debug\n",
    "\n",
    "        # different effect for strand\n",
    "            if STRAND == \"+\":\n",
    "\n",
    "            # different effect for the DELETION_SIZE\n",
    "                if DELETION_SIZE == \"small\":\n",
    "                    PORTION_TO_REMOVE = round(GENE_LENGHT/100*30)\n",
    "                    PORTION_TO_REMOVE_MINUS = GENE_LENGHT - PORTION_TO_REMOVE\n",
    "                elif DELETION_SIZE == \"big\":\n",
    "                    PORTION_TO_REMOVE = round(GENE_LENGHT/100*70)\n",
    "                    PORTION_TO_REMOVE_MINUS = GENE_LENGHT - PORTION_TO_REMOVE\n",
    "\n",
    "                gene_seq_removed = GENE_TO_REMOVE[:PORTION_TO_REMOVE]\n",
    "                gene_remove_index = seq.index(gene_seq_removed)\n",
    "                gene_seq_maintained = GENE_TO_REMOVE[PORTION_TO_REMOVE:]\n",
    "                #print(gene_seq_removed)\n",
    "                #print(\"maint: \"+gene_seq_maintained)\n",
    "\n",
    "        # different effect for strand\n",
    "            elif STRAND == \"-\":\n",
    "\n",
    "            # different effect for the DELETION_SIZE\n",
    "                if DELETION_SIZE == \"small\":\n",
    "                    PORTION_TO_REMOVE = round(GENE_LENGHT/100*30)\n",
    "                    PORTION_TO_REMOVE_MINUS = GENE_LENGHT - PORTION_TO_REMOVE\n",
    "                elif DELETION_SIZE == \"big\":\n",
    "                    PORTION_TO_REMOVE = round(GENE_LENGHT/100*70)\n",
    "                    PORTION_TO_REMOVE_MINUS = GENE_LENGHT - PORTION_TO_REMOVE\n",
    "\n",
    "                gene_seq_removed = GENE_TO_REMOVE[PORTION_TO_REMOVE_MINUS:]\n",
    "                gene_remove_index = seq.index(gene_seq_removed)\n",
    "                gene_seq_maintained = GENE_TO_REMOVE[:PORTION_TO_REMOVE_MINUS]\n",
    "                #print(gene_seq_removed)\n",
    "                #print(\"maint: \"+gene_seq_maintained)\n",
    "\n",
    "            seq_to_write += gene_seq_maintained\n",
    "            pos = pos+GENE_LENGHT-1\n",
    "            print(pos) # debug\n",
    "        #except:\n",
    "            #pass\n",
    "\n",
    "    for char in seq_to_write:\n",
    "        if len(seq_to_write_by_lines) == 80:\n",
    "            f.write(seq_to_write_by_lines+\"\\n\")\n",
    "            seq_to_write_by_lines = \"\"\n",
    "            pass\n",
    "        seq_to_write_by_lines += char\n",
    "    f.write(seq_to_write_by_lines+\"\\n\")\n",
    "    f.close()\n",
    "\n",
    "def translate_sequence(nucleotidic_seq):\n",
    "    '''\n",
    "    Translate genomic-retrieved sequence in AA alphabeth\n",
    "    --------------------------------------------------\n",
    "    INPUT:  nucleotidic sequence\n",
    "    OUTPUT: aminoacidic sequence\n",
    "    '''\n",
    "\n",
    "    aaseq = \"\"\n",
    "\n",
    "    t11 = {\"TTT\":\"F\", \"TTC\":\"F\", \"TTA\":\"L\", \"TTG\":\"L\",\n",
    "           \"TCT\":\"S\", \"TCC\":\"S\", \"TCA\":\"S\", \"TCG\":\"S\",\n",
    "           \"TAT\":\"Y\", \"TAC\":\"Y\", \"TAA\":\"*\", \"TAG\":\"*\",\n",
    "           \"TGT\":\"C\", \"TGC\":\"C\", \"TGA\":\"*\", \"TGG\":\"W\",\n",
    "           \"CTT\":\"L\", \"CTC\":\"L\", \"CTA\":\"L\", \"CTG\":\"L\",\n",
    "           \"CCT\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
    "           \"CAT\":\"H\", \"CAC\":\"H\", \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
    "           \"CGT\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\",\n",
    "           \"ATT\":\"I\", \"ATC\":\"I\", \"ATA\":\"I\", \"ATG\":\"M\",\n",
    "           \"ACT\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
    "           \"AAT\":\"N\", \"AAC\":\"N\", \"AAA\":\"K\", \"AAG\":\"K\",\n",
    "           \"AGT\":\"S\", \"AGC\":\"S\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
    "           \"GTT\":\"V\", \"GTC\":\"V\", \"GTA\":\"V\", \"GTG\":\"V\",\n",
    "           \"GCT\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\",\n",
    "           \"GAT\":\"D\", \"GAC\":\"D\", \"GAA\":\"E\", \"GAG\":\"E\",\n",
    "           \"GGT\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\",}\n",
    "\n",
    "    codons = [nucleotidic_seq[i:i+3] for i in range(0, len(nucleotidic_seq), 3)]\n",
    "    for el in codons:\n",
    "        if len(el) < 3:\n",
    "            codons.remove(el)\n",
    "\n",
    "    for codon in codons:\n",
    "        tr_codon = t11[codon]\n",
    "        aaseq += tr_codon\n",
    "\n",
    "    return aaseq\n",
    "\n",
    "def load_prodigal_genes(dir_prodigal, PRODIGAL_GENE_FILE):\n",
    "    genes = {}\n",
    "    \n",
    "    os.chdir(dir_prodigal)\n",
    "    with open(PRODIGAL_GENE_FILE) as f:\n",
    "        v = f.readlines()\n",
    "        f.seek(0)\n",
    "        start = -1\n",
    "        curr_h = \"\"\n",
    "        curr_f = \"\"\n",
    "        for line in v:\n",
    "            if line.startswith(\">\"):\n",
    "                genes[curr_h] = curr_f\n",
    "\n",
    "                curr_f = \"\"\n",
    "                curr_h = line.strip().split(\" \")[0]\n",
    "            else:\n",
    "                curr_f += line.strip()\n",
    "        genes[curr_h] = curr_f\n",
    "\n",
    "    genes.pop(\"\")\n",
    "    return genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# gene boundaries search (for gene removal purposes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "git_directory = os.getcwd()+\"/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_genomes = git_directory+\"5_complete_genomes/\"\n",
    "dir_prodigal = dir_genomes+\"prodigal/\"\n",
    "\n",
    "\n",
    "PRODIGAL_GENE_FILE = \"NZ_CP053946.1.fna\"\n",
    "MAG_TO_VIEW = \"NZ_CP053946.1.fna\"\n",
    "\n",
    "contigs, contigs_rev = load_MAG_and_divide(dir_genomes, MAG_TO_VIEW)\n",
    "\n",
    "GENE_OF_INTEREST = \">\"+\"NZ_CP053946.1_3603\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "genes = load_prodigal_genes(dir_prodigal, PRODIGAL_GENE_FILE)\n",
    "SEQ_gene_of_interest = genes[GENE_OF_INTEREST]\n",
    "LENGHT_gene_of_interest = len(SEQ_gene_of_interest)\n",
    "\n",
    "for fasta, seq in contigs.items():\n",
    "    if SEQ_gene_of_interest in seq:\n",
    "        bound_1 = seq.index(SEQ_gene_of_interest)\n",
    "        bound_2 = seq.index(SEQ_gene_of_interest)+LENGHT_gene_of_interest\n",
    "\n",
    "        #print(fasta+\" - genomic boundaries: \"+str(bound_1)+\"-\"+str(bound_2))\n",
    "        \n",
    "        #print(seq[bound_1:bound_2])\n",
    "\n",
    "for fasta, seq in contigs_rev.items():\n",
    "    if SEQ_gene_of_interest in seq:\n",
    "        lenght_contig = len(seq)\n",
    "\n",
    "        bound_1 = seq.index(SEQ_gene_of_interest)\n",
    "        bound_2 = seq.index(SEQ_gene_of_interest)+LENGHT_gene_of_interest\n",
    "\n",
    "        bound_2_plus = lenght_contig-bound_1\n",
    "        bound_1_plus = lenght_contig-bound_2\n",
    "        \n",
    "        bound_1 = bound_1_plus\n",
    "        bound_2 = bound_2_plus\n",
    "                \n",
    "        #print(fasta+\" (MINUS_STRAND)\"+\" - genomic boundaries: \"+str(bound_1)+\"-\"+str(bound_2))\n",
    "        \n",
    "        pre = \"ACTG\"\n",
    "        post = \"TGAC\"\n",
    "        compl = contigs[fasta].maketrans(pre,post)\n",
    "        fasta_compl = contigs[fasta].translate(compl)\n",
    "        rev_seq = fasta_compl[::-1]\n",
    "        \n",
    "        #print(contigs[fasta][bound_1:bound_2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## gene removal for MAG sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "LIST_GENES_TO_REMOVE = [\n",
    "    \">\",\n",
    "    \">\",\n",
    "    \">\"\n",
    "] # UPDATED EACH TIME WITH GENE NAMES FROM README.md\n",
    "\n",
    "dir_genomes = git_directory+\"0_control_group_results/\"\n",
    "dir_prodigal = dir_genomes+\"prodigal/\"\n",
    "\n",
    "PRODIGAL_GENE_FILE = \"NC_000913.3.fna\"\n",
    "MAG_TO_VIEW = \"NC_000913.3.fna\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "contigs, contigs_rev = load_MAG_and_divide(dir_genomes, MAG_TO_VIEW)\n",
    "genes = load_prodigal_genes(dir_prodigal, PRODIGAL_GENE_FILE)\n",
    "\n",
    "dict_of_genes_to_remove = {}\n",
    "\n",
    "for GENE in LIST_GENES_TO_REMOVE:\n",
    "    SEQ_gene_of_interest = genes[GENE]\n",
    "\n",
    "    LENGHT_gene_of_interest = len(SEQ_gene_of_interest)\n",
    "\n",
    "    n += 1\n",
    "\n",
    "    for fasta, seq in contigs.items():\n",
    "        if SEQ_gene_of_interest in seq:\n",
    "            bound_1 = seq.index(SEQ_gene_of_interest)\n",
    "            bound_2 = seq.index(SEQ_gene_of_interest)+LENGHT_gene_of_interest\n",
    "\n",
    "            dict_of_genes_to_remove.update({fasta:[str(bound_1),\"+\",seq[bound_1:bound_2]]})\n",
    "\n",
    "    for fasta, seq in contigs_rev.items():\n",
    "        if SEQ_gene_of_interest in seq:\n",
    "            lenght_contig = len(seq)\n",
    "\n",
    "            bound_1 = seq.index(SEQ_gene_of_interest)\n",
    "            bound_2 = seq.index(SEQ_gene_of_interest)+LENGHT_gene_of_interest\n",
    "\n",
    "            bound_2_plus = lenght_contig-bound_1\n",
    "            bound_1_plus = lenght_contig-bound_2\n",
    "\n",
    "            bound_1 = bound_1_plus\n",
    "            bound_2 = bound_2_plus\n",
    "\n",
    "            pre = \"ACTG\"\n",
    "            post = \"TGAC\"\n",
    "            compl = contigs[fasta].maketrans(pre,post)\n",
    "            fasta_compl = contigs[fasta].translate(compl)\n",
    "            rev_seq = fasta_compl[::-1]\n",
    "\n",
    "            dict_of_genes_to_remove.update({fasta:[str(bound_1),\"-\",contigs[fasta][bound_1:bound_2]]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "MAG_TO_REWRITE = MAG_TO_VIEW\n",
    "contigs = contigs\n",
    "\n",
    "#####################\n",
    "# TEST_1   bigINDELS#\n",
    "#####################\n",
    "dir_output = dir_genomes.replace(\"5_complete_genomes\", \"6_start_complete_bigINDELS\")\n",
    "DELETION_SIZE = \"big\"\n",
    "\n",
    "rewrite_MAG_removing_genes(MAG_TO_REWRITE, contigs, dir_output, DELETION_SIZE, dict_of_genes_to_remove)\n",
    "\n",
    "#####################\n",
    "# TEST_2 smallINDELS#\n",
    "#####################\n",
    "dir_output = dir_genomes.replace(\"5_complete_genomes\", \"7_start_complete_smallINDELS\")\n",
    "DELETION_SIZE = \"small\"\n",
    "\n",
    "rewrite_MAG_removing_genes(MAG_TO_REWRITE, contigs, dir_output, DELETION_SIZE, dict_of_genes_to_remove)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## removal for Genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "LIST_GENES_TO_REMOVE = [\n",
    "    \">\",\n",
    "    \">\",\n",
    "    \">\"\n",
    "] # UPDATED EACH TIME WITH GENE NAMES FROM README.md\n",
    "\n",
    "dir_genomes = git_directory+\"5_complete_genomes/\"\n",
    "dir_prodigal = dir_genomes+\"prodigal/\"\n",
    "\n",
    "PRODIGAL_GENE_FILE = \"NZ_CP053946.1.fna\"\n",
    "MAG_TO_VIEW = \"NZ_CP053946.1.fna\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "contigs, contigs_rev = load_MAG_and_divide(dir_genomes, MAG_TO_VIEW)\n",
    "genes = load_prodigal_genes(dir_prodigal, PRODIGAL_GENE_FILE)\n",
    "\n",
    "dict_of_genes_to_remove = {}\n",
    "\n",
    "n = -1\n",
    "for GENE in LIST_GENES_TO_REMOVE:\n",
    "    SEQ_gene_of_interest = genes[GENE]\n",
    "\n",
    "    LENGHT_gene_of_interest = len(SEQ_gene_of_interest)\n",
    "\n",
    "    n += 1\n",
    "\n",
    "    for fasta, seq in contigs.items():\n",
    "        if SEQ_gene_of_interest in seq:\n",
    "\n",
    "            bound_1 = seq.index(SEQ_gene_of_interest)\n",
    "            bound_2 = seq.index(SEQ_gene_of_interest)+LENGHT_gene_of_interest\n",
    "\n",
    "            dict_of_genes_to_remove.update({str(bound_1):[\"+\",seq[bound_1:bound_2]]})\n",
    "\n",
    "    for fasta, seq in contigs_rev.items():\n",
    "        if SEQ_gene_of_interest in seq:\n",
    "            lenght_contig = len(seq)\n",
    "\n",
    "            bound_1 = seq.index(SEQ_gene_of_interest)\n",
    "            bound_2 = seq.index(SEQ_gene_of_interest)+LENGHT_gene_of_interest\n",
    "\n",
    "            bound_2_plus = lenght_contig-bound_1\n",
    "            bound_1_plus = lenght_contig-bound_2\n",
    "\n",
    "            bound_1 = bound_1_plus\n",
    "            bound_2 = bound_2_plus\n",
    "\n",
    "            pre = \"ACTG\"\n",
    "            post = \"TGAC\"\n",
    "            compl = contigs[fasta].maketrans(pre,post)\n",
    "            fasta_compl = contigs[fasta].translate(compl)\n",
    "            rev_seq = fasta_compl[::-1]\n",
    "\n",
    "            dict_of_genes_to_remove.update({str(bound_1):[\"-\",contigs[fasta][bound_1:bound_2]]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENOME_TO_REWRITE = MAG_TO_VIEW\n",
    "contigs = contigs\n",
    "\n",
    "#####################\n",
    "# TEST_1   bigINDELS#\n",
    "#####################\n",
    "dir_output = dir_genomes.replace(\"5_complete_genomes\", \"6_start_complete_bigINDELS\")\n",
    "DELETION_SIZE = \"big\"\n",
    "\n",
    "rewrite_Genome_removing_genes(GENOME_TO_REWRITE, contigs, dir_output, DELETION_SIZE, dict_of_genes_to_remove)\n",
    "\n",
    "#####################\n",
    "# TEST_2 smallINDELS#\n",
    "#####################\n",
    "dir_output = dir_genomes.replace(\"5_complete_genomes\", \"7_start_complete_smallINDELS\")\n",
    "DELETION_SIZE = \"small\"\n",
    "\n",
    "rewrite_Genome_removing_genes(GENOME_TO_REWRITE, contigs, dir_output, DELETION_SIZE, dict_of_genes_to_remove)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# check MAG final results sequences - gene presence & BLASTp purposes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(git_directory+\"3_MAG_confrontation_and_final_results\")\n",
    "\n",
    "sequences = {}\n",
    "\n",
    "with open(\"MAG_FINAL_RESULTS.tsv\") as f:\n",
    "    for line in f.readlines()[2:]:\n",
    "        line = line.strip().split(\"\\t\")\n",
    "\n",
    "        MAG = line[0]\n",
    "        KO = line[1]\n",
    "        seq_0 = line[2]\n",
    "        xseq_0 = line[3]\n",
    "        seq_1 = line[4]\n",
    "        xseq_1 = line[5]\n",
    "        seq_2 = line[6]\n",
    "        xseq_2 = line[7]\n",
    "        \n",
    "        if not seq_0 == \"NaN\":\n",
    "            fasta_0 = \">\"+MAG+\"_\"+KO+\"__0\"\n",
    "            sequences.update({fasta_0:seq_0})\n",
    "        if not seq_1 == \"NaN\":\n",
    "            fasta_1 = \">\"+MAG+\"_\"+KO+\"__1\"\n",
    "            sequences.update({fasta_1:seq_1})\n",
    "        if not seq_2 == \"NaN\":\n",
    "            fasta_2 = \">\"+MAG+\"_\"+KO+\"__2\"\n",
    "            sequences.update({fasta_2:seq_2})\n",
    "\n",
    "#print(sequences)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prodigal gene presence/lack thereof"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# code to generate \"tests/3_MAG_confrontation_and_final_results/other_data/hits_found_in_0_prodigal.txt\"\n",
    "dir_prodigal = git_directory+\"0_control_group_results/prodigal/\"\n",
    "\n",
    "for el in os.listdir(dir_prodigal):\n",
    "    if el.endswith(\".fna\"):\n",
    "        PRODIGAL_GENE_FILE = el\n",
    "        print(\"++\"+PRODIGAL_GENE_FILE+\"\\n\")\n",
    "        try:\n",
    "            genes = load_prodigal_genes(dir_prodigal, PRODIGAL_GENE_FILE)\n",
    "\n",
    "            for gene in genes:\n",
    "\n",
    "                for header, hmmhit_seq in sequences.items():\n",
    "                    if hmmhit_seq in genes[gene]:\n",
    "                        print(header)\n",
    "                        print(\"\\t\"+gene) # to check if the same \n",
    "            print(\"\")\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# code to generate \"tests/3_MAG_confrontation_and_final_results/other_data/hits_found_in_2_prodigal.txt\"\n",
    "dir_prodigal = git_directory+\"2_smallINDELS_group_results/prodigal/\"\n",
    "\n",
    "for el in os.listdir(dir_prodigal):\n",
    "    if el.endswith(\".fna\"):\n",
    "        PRODIGAL_GENE_FILE = el\n",
    "        print(\"++\"+PRODIGAL_GENE_FILE+\"\\n\")\n",
    "        try:\n",
    "            genes = load_prodigal_genes(dir_prodigal, PRODIGAL_GENE_FILE)\n",
    "\n",
    "            for gene in genes:\n",
    "\n",
    "                for header, hmmhit_seq in sequences.items():\n",
    "                    if hmmhit_seq in genes[gene]:\n",
    "                        print(header)\n",
    "                        print(\"\\t\"+gene) # to check if the same \n",
    "            print(\"\")\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# code to generate \"tests/3_MAG_confrontation_and_final_results/other_data/hits_found_in_1_prodigal.txt\"\n",
    "dir_prodigal = git_directory+\"1_bigINDELS_group_results/prodigal/\"\n",
    "\n",
    "for el in os.listdir(dir_prodigal):\n",
    "    if el.endswith(\".fna\"):\n",
    "        PRODIGAL_GENE_FILE = el\n",
    "        print(\"++\"+PRODIGAL_GENE_FILE+\"\\n\")\n",
    "        try:\n",
    "            genes = load_prodigal_genes(dir_prodigal, PRODIGAL_GENE_FILE)\n",
    "\n",
    "            for gene in genes:\n",
    "\n",
    "                for header, hmmhit_seq in sequences.items():\n",
    "                    if hmmhit_seq in genes[gene]:\n",
    "                        print(header)\n",
    "                        print(\"\\t\"+gene) # to check if the same \n",
    "            print(\"\")\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# check Complete genomes final results sequences - gene presence & BLASTp purposes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(git_directory+\"7_COMPLETE_GENOMES_confrontation_and_final_results\")\n",
    "\n",
    "sequences = {}\n",
    "\n",
    "with open(\"COMPLETE_FINAL_RESULTS.tsv\") as f:\n",
    "    for line in f.readlines()[2:]:\n",
    "        line = line.strip().split(\"\\t\")\n",
    "\n",
    "        GENOME = line[0]\n",
    "        KO = line[1]\n",
    "        seq_0 = line[2]\n",
    "        xseq_0 = line[3]\n",
    "        seq_1 = line[4]\n",
    "        xseq_1 = line[5]\n",
    "        seq_2 = line[6]\n",
    "        xseq_2 = line[7]\n",
    "        \n",
    "        if not seq_0 == \"NaN\":\n",
    "            fasta_0 = \">\"+GENOME+\"_\"+KO+\"__0\"\n",
    "            sequences.update({fasta_0:seq_0})\n",
    "        if not seq_1 == \"NaN\":\n",
    "            fasta_1 = \">\"+GENOME+\"_\"+KO+\"__1\"\n",
    "            sequences.update({fasta_1:seq_1})\n",
    "        if not seq_2 == \"NaN\":\n",
    "            fasta_2 = \">\"+GENOME+\"_\"+KO+\"__2\"\n",
    "            sequences.update({fasta_2:seq_2})\n",
    "\n",
    "#print(sequences)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prodigal gene presence/lack thereof"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# code to generate \"tests/7_COMPLETE_GENOMES_confrontation_and_final_results/other_data/hits_found_in_4_prodigal.txt\"\n",
    "dir_prodigal = git_directory+\"4_complete_control_group_results/prodigal/\"\n",
    "\n",
    "for el in os.listdir(dir_prodigal):\n",
    "    if el.endswith(\".fna\"):\n",
    "        PRODIGAL_GENE_FILE = el\n",
    "        print(\"++\"+PRODIGAL_GENE_FILE+\"\\n\")\n",
    "        try:\n",
    "            genes = load_prodigal_genes(dir_prodigal, PRODIGAL_GENE_FILE)\n",
    "\n",
    "            for gene in genes:\n",
    "\n",
    "                for header, hmmhit_seq in sequences.items():\n",
    "                    if hmmhit_seq in genes[gene]:\n",
    "                        print(header)\n",
    "                        print(\"\\t\"+gene) # to check if the same \n",
    "            print(\"\")\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# code to generate \"tests/7_COMPLETE_GENOMES_confrontation_and_final_results/other_data/hits_found_in_6_prodigal.txt\"\n",
    "dir_prodigal = git_directory+\"6_complete_smallINDELS_group_results/prodigal/\"\n",
    "\n",
    "for el in os.listdir(dir_prodigal):\n",
    "    if el.endswith(\".fna\"):\n",
    "        PRODIGAL_GENE_FILE = el\n",
    "        print(\"++\"+PRODIGAL_GENE_FILE+\"\\n\")\n",
    "        try:\n",
    "            genes = load_prodigal_genes(dir_prodigal, PRODIGAL_GENE_FILE)\n",
    "\n",
    "            for gene in genes:\n",
    "\n",
    "                for header, hmmhit_seq in sequences.items():\n",
    "                    if hmmhit_seq in genes[gene]:\n",
    "                        print(header)\n",
    "                        print(\"\\t\"+gene) # to check if the same \n",
    "            print(\"\")\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# code to generate \"tests/7_COMPLETE_GENOMES_confrontation_and_final_results/other_data/hits_found_in_5_prodigal.txt\"\n",
    "dir_prodigal = git_directory+\"5_complete_bigINDELS_group_results/prodigal/\"\n",
    "\n",
    "for el in os.listdir(dir_prodigal):\n",
    "    if el.endswith(\".fna\"):\n",
    "        PRODIGAL_GENE_FILE = el\n",
    "        print(\"++\"+PRODIGAL_GENE_FILE+\"\\n\")\n",
    "        try:\n",
    "            genes = load_prodigal_genes(dir_prodigal, PRODIGAL_GENE_FILE)\n",
    "\n",
    "            for gene in genes:\n",
    "\n",
    "                for header, hmmhit_seq in sequences.items():\n",
    "                    if hmmhit_seq in genes[gene]:\n",
    "                        print(header)\n",
    "                        print(\"\\t\"+gene) # to check if the same \n",
    "            print(\"\")\n",
    "        except:\n",
    "            pass"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
